lib/seahorn/PredicateAbstraction.cc

```diff
@@ -63,22 +66,72 @@ namespace seahorn
 		boost::tribool result = fp.query ();
 //
 		if (result) outs () << "sat";
-		else if (!result) outs () << "unsat";
+		else if (!result)
+		{
+			outs() << "unsat\n";
+			HornDbModel absModel(new_db, fp);
+			converter.convert(absModel, oldModel, newToOldPredMap);
+			for(ExprMap::iterator it = oldModel.getRelToSolutionMap().begin(); it!=oldModel.getRelToSolutionMap().end(); ++it)
+			{
+				outs() << *(bind::fname(it->first)) << ": " << *(it->second) << "\n";
+			}
+		}
 		else outs () << "unknown";
 		outs () << "\n";
-//
-		printInvars (M);
 
 		return false;
 	}
 
+	bool HornModelConverter::convert (HornDbModel &in, HornDbModel &out, std::map<Expr, Expr> &newToOldPredMap)
+	{
+		ExprMap inRelToSolMap = in.getRelToSolutionMap();
```

Comment:
Sorry I'm a little confused about this comment. In the convert function we want to convert the abstract relation to the original relation, and convert the boolean variables to the terms. It seems we need such a data structure for iterating over the relations in abstractDB, no matter what the internal implementation is. One alternative design that I can think out is not to return the map, but a list of abs_relations, and for each relation, I use a method such as HornDbModel.getBv(rel) to get its boolean variable. Is this reasonable?